pycalphad/io/cs_dat.py

```diff
@@ -1159,6 +1162,999 @@ def parse_cs_dat(instring):
+            detect_query = (
+                (where("phase_name") == phase_name) & \
+                (where("parameter_type") == "TC")
+            )
+            tcs = dbf._parameters.search(detect_query)
+
+            detect_query = (
+                (where("phase_name") == phase_name) & \
+                (where("parameter_type") == "BMAGN")
+            )
+            bmagns = dbf._parameters.search(detect_query)
+
+        # Get endmembers and other parameters depending on phase model
+        if phase_model in ('SUBG', 'SUBQ'):
+            # Get parameters for endmembers
+            detect_query = (
+                (where("phase_name") == phase_name) & \
+                (where("parameter_type") == "MQMG")
+            )
+            endmember_params = dbf._parameters.search(detect_query)
+            # Write zeta for SUBG
+            if phase_model == 'SUBG':
+                # All zetas are the same, so grab the first one
+                zeta = endmember_params[0]['zeta']
+                output += f'{zeta:9.5f}\n'
+            # Get parameters for quadruplets
+            detect_query = (
+                (where("phase_name") == phase_name) & \
+                (where("parameter_type") == "MQMZ")
+            )
+            quadruplet_params = dbf._parameters.search(detect_query)
+            # Write number of endmembers and number of non-default coordination sets next
+            number_of_endmembers = len(endmember_params)
+            number_of_non_default_quadruplets = len(quadruplet_params)
+            output += f'{number_of_endmembers:4} {number_of_non_default_quadruplets:3}\n'
+        elif phase_model in ('IDMX', 'RKMP', 'RKMPM', 'QKTO', 'SUBL', 'SUBLM', 'PITZ'):
+            # Get species for CEF phases
+            detect_query = (
+                (where("phase_name") == phase_name) & \
+                (where("parameter_type") == "G")
+            )
+            endmember_params = []
+            excess_params = []
+            for endmember in dbf._parameters.search(detect_query):
+                is_endmember = all([len(subl) == 1 for subl in endmember['constituent_array']])
+                if is_endmember:
+                    endmember_params.append(endmember)
+                else:
+                    excess_params.append(endmember)
+
+
+        # Get sublattice weights
+        if phase_model in ('SUBL','SUBLM'):
+            sublattice_weights = itertools.cycle(dbf.phases[phase_name].sublattices)
+        else:
+            sublattice_weights = itertools.cycle([1])
+
+        # Write endmember data
+        endmember_names = []
+        for endmember in endmember_params:
+            # Generate species names and stoichiometries
+            name = ''
+            stoichiometry = [0 for _ in elements_ordered]
+            # Overwrite sublattice_weights for MQM phases
+            if phase_model in ('SUBG', 'SUBQ'):
+                sublattice_weights = itertools.cycle(endmember['stoichiometry'])
```

Comment:
`cycle` is convenient, but no doubt is not the only option. For some phase models we have sublattices 1 through N each with a weight, and need to iterate through them, in order, repeatedly. For others the weights are all 1. This seemed to capture both cases neatly.

Is there any chance of the sublattice order getting shuffled? This would break if some endmembers have their constituents listed out of order.